examples/dynamoathenamigration/migration.go

```diff
@@ -486,18 +489,251 @@ func (t *task) downloadFromS3(ctx context.Context, key string) (*os.File, error)
+	// will be loaded into memory, sorted and stored in temporary files.
+	maxMemoryUsedForSortingExportInBytes := 1024 * 1024 * t.MaxMemoryUsedForSortingExportInMB
+	sortedFile, err := createSortedExport(originalFile, dir, maxMemoryUsedForSortingExportInBytes)
+	if err != nil {
+		return nil, trace.NewAggregate(err, sortedFile.Close())
+	}
+	return sortedFile, nil
+}
+
+type eventWithTime struct {
+	rawLine   json.RawMessage
+	eventDate string
+}
+
+// createSortedExport reads a large json.gz export file, which cannot fit entirely
+// into memory and splits it into multiple smaller sorted (in memory) files.
+// Afterwards, it merges these smaller files into a final sorted output.
+// The function returns json file which is sorted by createdAtDate ascending.
+func createSortedExport(inFile *os.File, dir string, maxSize int) (f *os.File, err error) {
+	tmpSortedFilesDir := path.Join(dir, fmt.Sprintf("sort-%s", uuid.NewString()))
+	if err := os.Mkdir(tmpSortedFilesDir, 0o755); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer func() {
+		err = trace.NewAggregate(err, os.RemoveAll(tmpSortedFilesDir))
+	}()
+
+	gzipReader, err := gzip.NewReader(inFile)
+	if err != nil {
 		return nil, trace.Wrap(err)
 	}
-	return f, nil
+	defer gzipReader.Close()
+
+	dec := json.NewDecoder(gzipReader)
+	events := make([]eventWithTime, 0, 10000)
+	var tmpFiles []string
+	var size int
+	// dec.More read export line by line because export is new line json.
+	for dec.More() {
+		// First decode whole line to rawMessage.
+		var singleLine json.RawMessage
+		err = dec.Decode(&singleLine)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		// Parse just createAtDate.
+		var parsedLine dynamoEventPart
+		err = json.Unmarshal(singleLine, &parsedLine)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		events = append(events, eventWithTime{
+			rawLine:   singleLine,
+			eventDate: parsedLine.Item.CreatedAtDate.Value,
+		})
+		size += len(singleLine)
+		if size >= maxSize {
+			// When max size is reached, sort events and write tmp file.
+			fname, err := sortEventsAndWriteTmpFile(tmpSortedFilesDir, events)
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+			tmpFiles = append(tmpFiles, fname)
+			// clear size and events slice, keeping the buffer.
+			events = events[:0]
+			size = 0
+		}
+	}
+	// last batch, after last event was read from file.
+	if size > 0 {
+		fname, err := sortEventsAndWriteTmpFile(tmpSortedFilesDir, events)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		tmpFiles = append(tmpFiles, fname)
+	}
+	finalFile := path.Join(dir, fmt.Sprintf("final_sorted_%s.json", uuid.NewString()))
+	err = mergeFiles(tmpFiles, finalFile)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	f, err = os.Open(finalFile)
+	return f, err
+}
+
+func sortEventsAndWriteTmpFile(dir string, events []eventWithTime) (string, error) {
+	sort.SliceStable(events, func(i, j int) bool {
```

Comment:
Maybe swap places sorting and creating temp file? Just so we don't do sorting for nothing if we can't create a file.

Reply:
https://github.com/gravitational/teleport/pull/29627/commits/bc585a406c93ee5d7fddb198b0d1b46f983ca737